Add systemctl list-units mode to print slice tree structure

devdekunle opened the pull request:
Implemented this task in the TODO which enhance systemctl list-units by adding the  option --by-slice" mode, that rearranges the
output of "systemctl list-units" slightly by showing the tree structure of
 the slices, and the units attached to them.

devdekunle commented:
Attached sample output
![Screenshot from 2024-03-30 02-43-38](https://github.com/systemd/systemd/assets/110066922/e90291e4-bd7f-4e98-aaf8-8955566df194)

devdekunle commented:
Attached sample output
![Screenshot from 2024-03-30 02-49-52](https://github.com/systemd/systemd/assets/110066922/126f2ff2-7a00-4684-9ae0-e81fbdcbc51a)

devdekunle commented:
sample output
![Screenshot from 2024-03-30 02-51-16](https://github.com/systemd/systemd/assets/110066922/f1bed5cd-01cb-4d24-b4b3-147024ddd608)

devdekunle commented:
sample output
![Screenshot from 2024-03-30 02-52-02](https://github.com/systemd/systemd/assets/110066922/d6dc9a81-cbf4-4942-8c1d-7985ac07911d)

devdekunle commented:
Hello @bluca. Please kindly review a long task

devdekunle commented:
Hello @bluca , I hope you are good sir. Please I wanted to know if i should write unittest or integration test for this after you have reviewed. Thanks you as i await your response.

devdekunle commented:
Hello sir @bluca. Good afternoon sir and I hope you are good.
Please can you review sir?

bluca commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +107,205 @@ static char *format_unit_id(const char *unit, const char *machine) {
+} MapNode;
+
+/* Hash map definition */
+MapNode *hash_map[MAP_SIZE] = {NULL};
```
do not use global variables for this

bluca commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +107,205 @@ static char *format_unit_id(const char *unit, const char *machine) {
+MapNode *hash_map[MAP_SIZE] = {NULL};
+
+/* Function to calculate hash value of the slice name */
+static size_t hash(const char *slice_name) {
```
do not reinvent hashing, use the available src/basic/hashmap

bluca commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +107,205 @@ static char *format_unit_id(const char *unit, const char *machine) {
+        for(size_t i = 0; i < MAP_SIZE; i++) {
+              current =  hash_map[i];
+              while (current != NULL) {
+                next = current->hashmap_next;
```
indentation and formatting are wrong, look at docs/CODING_STYLE.md and adjust here and elsewhere

bluca commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +107,205 @@ static char *format_unit_id(const char *unit, const char *machine) {
         int r;
 
+        if (arg_list_by_slice) {
+                FOREACH_ARRAY(u, unit_infos, c) {
```
move this to a separate function

bluca reviewed (changes requested):
Add documentation in the manpage and test coverage in test/units/testsuite-26.sh

devdekunle commented:
Okay thank you very much sir @bluca for the review.

devdekunle commented:
> Add documentation in the manpage and test coverage in test/units/testsuite-26.sh

I am not sure about this sir. I tried running the tests locally but it get some errors when I run
Sudo test/run-integration-tests.sh or when I try to run just the systemctl test directory
Sudo make -C test/TEST-26-SYSTEMCTL clean setup run. 
I get some errors such as could not find Qemu binary and also /home/.../systemd/build needs to be built with -Dinstall-test=true. 
I tried setting up Qemu on mkosi but I've be encountering an error when trying to generate the image. Please can you kindly help me sir.

bluca commented:
Follow the instructions that the error gives you - install qemu and configure meson with -Dinstall-tests=true

devdekunle commented:
> Follow the instructions that the error gives you - install qemu and configure meson with -Dinstall-tests=true

`‣  Generating unified kernel image for kernel version 6.8.0-11-generic
/usr/lib/systemd/ukify: 1: exec: /usr/lib/systemd/ukify: not found
‣ "/usr/lib/systemd/ukify --cmdline @/home/adekunle/.cache/mkosi/mkosi-workspace0guekfhd/cmdline --os-release @/home/adekunle/.cache/mkosi/mkosi-workspace0guekfhd/root/usr/lib/os-release --stub /home/adekunle/.cache/mkosi/mkosi-workspace0guekfhd/root/usr/lib/systemd/boot/efi/linuxx64.efi.stub --output /home/adekunle/.cache/mkosi/mkosi-workspace0guekfhd/root/boot/EFI/Linux/ubuntu-6.8.0-11-generic.efi --efi-arch x64 --uname 6.8.0-11-generic --sign-kernel --signtool sbsign --secureboot-private-key /home/adekunle/systemd/mkosi.key --secureboot-certificate /home/adekunle/systemd/mkosi.crt --pcr-private-key /home/adekunle/systemd/mkosi.key --pcr-banks sha1,sha256 build --linux /home/adekunle/.cache/mkosi/mkosi-workspace0guekfhd/root/usr/lib/modules/6.8.0-11-generic/vmlinuz --initrd /home/adekunle/.cache/mkosi/mkosi-workspace0guekfhd/microcode.initrd --initrd /home/adekunle/.cache/mkosi/mkosi-workspace0guekfhd/initrd --initrd /home/adekunle/.cache/mkosi/mkosi-workspace0guekfhd/kernel-modules-6.8.0-11-generic.initrd" returned non-zero exit code 127.`
I get this error during the installation process when trying to setup qemu
sudo mkosi -f qemu.
At some point during the installation before the seutp exited as mentioned above, a message saying 
`System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down
`.
Please how do i go about this sir @bluca.

devdekunle commented:
@bluca sir, going through the failed tests of the CentosOS CI i saw where the failure is coming from in the test coverage.
`11:02:22 [    9.167863] testsuite-26.sh[383]: + systemctl list-units --all --by-slice
11:02:22 [    9.176832] systemctl[571]: Assertion 'units && num_units > 0' failed at src/systemctl/systemctl-list-units.c:121, function insert_units_under_slice(). Aborting.
11:02:22 [    9.470658] systemd-coredump[573]: Process 571 (systemctl) of user 0 dumped core.
11:02:22                                       
11:02:22                                       Stack trace of thread 571:
11:02:22                                       #0  0x00007cafd0aab32c n/a (libc.so.6 + 0x8d32c)
11:02:22                                       #1  0x00007cafd0a5a6c8 raise (libc.so.6 + 0x3c6c8)
11:02:22                                       #2  0x00007cafd0a424b8 abort (libc.so.6 + 0x244b8)
11:02:22                                       #3  0x00007cafd0e1dcb7 log_assert_failed (libsystemd-shared-256.so + 0x21dcb7)
11:02:22                                       #4  0x000060144491f31c insert_units_under_slice (systemctl + 0x1731c)
11:02:22                                       #5  0x000060144491f698 fetch_bus_property (systemctl + 0x17698)
11:02:22                                       #6  0x000060144491feed output_units_list (systemctl + 0x17eed)
11:02:22                                       #7  0x0000601444923567 verb_list_units (systemctl + 0x1b567)
11:02:22                                       #8  0x00007cafd0dd6be3 dispatch_verb (libsystemd-shared-256.so + 0x1d6be3)
11:02:22                                       #9  0x000060144493e8b5 systemctl_main (systemctl + 0x368b5)
11:02:22                                       #10 0x000060144493ee60 run (systemctl + 0x36e60)
11:02:22                                       #11 0x000060144493eef8 main (systemctl + 0x36ef8)
11:02:22                                       #12 0x00007cafd0a43cd0 n/a (libc.so.6 + 0x25cd0)
11:02:22                                       #13 0x00007cafd0a43d8a __libc_start_main (libc.so.6 + 0x25d8a)
11:02:22                                       #14 0x0000601444914775 _start (systemctl + 0xc775)
11:02:22                                       ELF object binary architecture: AMD x86-64
11:02:22 [    9.485888] testsuite-26.sh[383]: /usr/lib/systemd/tests/testdata/units/testsuite-26.sh: line 104:   571 Aborted                 (core dumped) systemctl list-units --all --by-slice
11:02:22 [    9.485888] testsuite-26.sh[383]: + at_exit
11:02:22 [    9.485888] testsuite-26.sh[383]: + [[ -v UNIT_NAME ]]
11:02:22 [    9.485888] testsuite-26.sh[383]: + [[ -e /usr/lib/systemd/system/systemctl-test-25789.service ]]
11:02:22 [    9.485888] testsuite-26.sh[383]: + rm -fvr /usr/lib/systemd/system/systemctl-test-25789.service /etc/systemd/system/systemctl-test-25789.service.d +4
11:02:22 [    9.530995] testsuite-26.sh[575]: removed '/usr/lib/systemd/system/systemctl-test-25789.service'
11:02:22 [    9.530995] testsuite-26.sh[575]: removed '/etc/systemd/system/systemctl-test-25789.service.d/override.conf'
11:02:22 [    9.530995] testsuite-26.sh[575]: removed '/etc/systemd/system/systemctl-test-25789.service.d/override2.conf'
11:02:22 [    9.530995] testsuite-26.sh[575]: removed directory '/etc/systemd/system/systemctl-test-25789.service.d'
11:02:22 [    9.531170] testsuite-26.sh[383]: + maybe_umount_usr_overlay
11:02:22 [    9.531170] testsuite-26.sh[383]: + false
11:02:22 [    9.531170] testsuite-26.sh[383]: + rm -f /etc/init.d/issue-24990
11:02:22 [    9.531170] testsuite-26.sh[383]: + return 0
11:02:22 [    9.531865] systemd[1]: testsuite-26.service: Failed with result 'exit-code'.
11:02:22 [    9.531966] systemd[1]: Failed to start testsuite-26.service.
11:02:22 Spawning getter /systemd-meson-build/journalctl -o export -D /var/tmp/systemd-test-TEST-26-SYSTEMCTL/root/var/log/journal...
11:02:22 Creating journal file /var/tmp/systemd-test-TEST-26-SYSTEMCTL/system.journal on a btrfs file system, and copy-on-write is enabled. This is likely to slow down journal access substantially, please consider turning off the copy-on-write file attribute on the journal directory, using chattr +C.
11:02:22 Finishing after writing 28598 entries
11:02:22 TEST-26-SYSTEMCTL: (failed; see logs)
11:02:22 -rw-r----- 1 root root 25165824 Apr  4 10:02 /var/tmp/systemd-test-TEST-26-SYSTEMCTL/system.journal
11:02:22 qemu test failed
11:02:22 make: *** [Makefile:4: run] Error 1
11:02:22 make: Leaving directory '/build/test/TEST-26-SYSTEMCTL'
11:02:22 Command exited with non-zero status 2`
The second line of this failure above showed the assertion failure of the units and num_units gotten from bus_get_property_strv() function. 
`11:02:22 [    9.176832] systemctl[571]: Assertion 'units && num_units > 0' failed at src/systemctl/systemctl-list-units.c:121, function insert_units_under_slice(). Aborting.`
I experienced this failure locally after running `sudo make -C test/TEST-26-SYSTEMCTL clean setup run`.
Before i did this test run, the code works perfectly but I get that assertion error after i run it. from my debugging, the dbus doesn't return the properties hence the error. This occurs after i run the `sudo make -C test/TEST-26-SYSTEMCTL clean setup run` command or `sudo test/run-integration-tests.sh`. 
is this failure because of qemu not working yet sir?

devdekunle commented:
Hello @bluca @DaanDeMeyer @yuwata @poettering Please sirs I am currently having issues running my integration tests and setting up mkosi and Qemu. 
After adding my tests to the test/units/testsuite-26.sh, I try running just that test suite using
`sudo make -C test/TEST-26-SYSTEMCTL clean setup run` but I receive an error saying 
`build/ needs to be configures with -Dinstall-test=true`. I have cleared the initial configuration and set up meson with his option but I still get the same error. 
Also when I try to run all the integration tests locally, using `sudo test/run-integration-tests.sh`. I get some errors saying `qemu binary not found` although that didn't stop the setup process, but along the line, It fails and all the tests fails with the same -Dinstall-test=true error. Hence I try to setup mkosi but it also fails at some point with this error `exec /usr/lib/systemd/ukify not found` and also at some point before getting to this, it also brings a message which says `System has not been booted with systemd as init system (PID 1). Can't operate. Failed to connect to bus: Host is down `. But when i run `systemctl list-units`, init.scope it shows the unit as loaded and active. Please 🙏🏽 I need assistance sirs.

bluca commented:
You need to install qemu, search on google for instruction on how to do that depending on your distribution

bluca commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,200 @@ static char *format_unit_id(const char *unit, const char *machine) {
 }
 
-static int output_units_list(const UnitInfo *unit_infos, size_t c) {
+static Hashmap *slice_unit_map = NULL;
```
do not use global variables for this

bluca commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,200 @@ static char *format_unit_id(const char *unit, const char *machine) {
+        assert(units && num_units > 0);
+        /* Initialize the hashmap */
+        if (!slice_unit_map)
+                slice_unit_map = hashmap_new(&string_hash_ops);
```
OOM check missing

bluca commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,200 @@ static char *format_unit_id(const char *unit, const char *machine) {
+typedef struct {
+        char ** units;
+        size_t num_units;
+}SliceUnits;
```
just use a strv, see src/basic/strv.h

devdekunle replied:
Thank you for your review sir @bluca, for this case i used a struct instead of strv because both the `hashmap_put` and `hashmap_get` both take a pointer and return a pointer respectively. 
`static inline void *hashmap_get(Hashmap *h, const void *key)` and 
`static inline int hashmap_put(Hashmap *h, const void *key, void *value) `
However the units under a slice is an array of pointers to strings(char **). so that is why i used the struct to just pass the functions its pointer.

devdekunle commented:
> You need to install qemu, search on google for instruction on how to do that depending on your distribution

Thank you very much @bluca , I was able to test with mkosi and systemd-nspawn.

devdekunle commented:
Hello @bluca, Good morning sir. I just wanted to remind you of the review. and also, please what do you think about this two issues. I would love to implement them.
1.#29180
2. #32137.
Thank you very much for your guidance and mentorship so far.

poettering commented on man/systemctl.xml:
```diff
@@ -64,7 +64,9 @@
-
+            <para>
+              The <option>--by-slice</option> shows the tree structure of slices for the units filtered by <option>--type</option>, <option>--state</option> or <option>--all</option>. For the tree structure to be shown, loaded units must have the root slice '-.slice'. Additionally, units can be passed as argument without any of the options stated above. For such cases, the root slice must also be ...
+            </para>
```
please follow the usual style around here, i.e. no newlines after `<para>`

poettering replied:
please line break at 109 chars

poettering replied:
`-.slice` should be enclosed in `<literal></literal>`

poettering commented on man/systemctl.xml:
```diff
@@ -64,7 +64,9 @@
             options are specified.</para>
-
+            <para>
+              The <option>--by-slice</option> shows the tree structure of slices for the units filtered by <option>--type</option>, <option>--state</option> or <option>--all</option>. For the tree structure to be shown, loaded units must have the root slice '-.slice'. Additionally, units can be passed as argument without any of the options stated above. For such cases, the root slice must also be ...
```
we have this rule that switches that take parameters should always be suffixed with a `=` in our docs. i.e. `--type=`, `--state=`

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
-static int output_units_list(const UnitInfo *unit_infos, size_t c) {
+typedef struct {
+        char ** units;
+        size_t num_units;
```
just make this an strv, i.e. a NULL terminated string array, see strv.h

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+typedef struct {
+        char ** units;
+        size_t num_units;
+}SliceUnits;
```
space missing between `}` and `SliceUnits`

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+                        return log_oom();
+        }
+        /* Allocate memory for array of unit names for a slice */
+        slice_units = new(SliceUnits, sizeof(SliceUnits));
```
2nd arg should be the number of objects of the type specified as 1st arg.

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+                slice_unit_map = hashmap_new(&string_hash_ops);
+                if (!slice_unit_map)
+                        return log_oom();
+        }
```
please don't allocate ahead of time, but do this on first insertion. i.e. hashmap_ensure_put() makes that easy.

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+        /* Allocate memory for array of unit names for a slice */
+        slice_units = new(SliceUnits, sizeof(SliceUnits));
+        if (!slice_units) {
+                mfree(slice_units);
```
please use `_cleanup_free_` macro

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+                return log_oom();
+        }
+
+        slice_units->units = new(char *, sizeof(char *) * num_units);
```
this is not how the `new()` macro is used. first arg type, 2nd arg is number of objects of that type

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+                for (size_t i = 0; i < num_units; i++)
+                        mfree(slice_units->units[i]);
+                mfree(slice_units->units);
+                mfree(slice_units);
```
do not use `mfree()` btw unless you want to use its return value. if you don#t just use `free()`.

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+                mfree(slice_units);
+                return log_oom();
+        }
+        return r;
```
seriousuly, just use an strv() here

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+
+        assert(slice_name);
+        if (!slice_unit_map)
+                return NULL;
```
hashmap_get() is happy if you pass NULL as first arg. It treats that like an empty hashmap

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+
+        hashmap_free(slice_unit_map);
+        }
+}
```
indentation is borked.

poettering commented on src/systemctl/systemctl.h:
```diff
@@ -104,6 +104,8 @@ extern bool arg_mkdir;
 extern const char *arg_drop_in;
 extern ImagePolicy *arg_image_policy;
+extern bool arg_list_by_slice;
+extern Hashmap *slice_unit_map;
```
nope, we generally do not want global variables in our codebase, with some very specific exceptions, i.e. if the variables are directly initialized from data parsed from inherently process global concepts, such as env vars and command line options.

For the rest, please pass the variables through and define them with a clear context

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -537,6 +540,7 @@ static int systemctl_parse_argv(int argc, char *argv[]) {
                 { "drop-in",             required_argument, NULL, ARG_DROP_IN             },
                 { "when",                required_argument, NULL, ARG_WHEN                },
                 { "stdin",               no_argument,       NULL, ARG_STDIN               },
+                { "by-slice",              no_argument,     NULL, ARG_LIST_BY_SLICE               },
```
please align this line's fields like the others

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+        if (!arg_has_root_slice) {
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                        "No root slice in loaded units.\nLoaded units must have root slice '-.slice' to print slice tree structure.\n");
+        }
```
no {} around single-line if blocks

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+                        arg_has_root_slice = true;
+                        break;
+                }
+                continue;
```
redundant line

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+
+        for(int i = 0; i < level; ++i)
+                printf("|   ");
+        printf("|__ %s\n", slice_name);
```
please use the nice unicode chars for this accessible in special_glyph()

poettering commented on src/systemctl/systemctl-list-units.c:
```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+                        printf("|___%s\n", units[i]);
+                }
+        }
+        return 0;
```
instead of outputting something completely new, just rearrange the usual tabular output that output_units_list() generates.

For that, please add an additional columns to the table_new() call in output_units_list(): one that contains the slice the unit is in. This should be populated for each unit, but be hidden table_hide_column_from_display(). In --by-slice mode it should be used as the primary sort key however, see table_set_sort(). 

And then, simply change the unit name column so that it is prefixed with the tree branches as appropriate, if --by-slice mode is on.

Thus, we get the usual tabular output, the --by-slice option just adjusts the ordering to take slices into account, and draws the tree structure in front of the unit name

devdekunle replied:
Hello sir @poettering , I have been having some challenges implementing the table output this way, I have been able to add the slice row for each unit and also added the slice each unit belongs to in the slice column. However i have been have challenges as regards the sorting and how it positions the units on the table which prevents the branches from being prefixed appropriately. I tried sorting table_set_sort() just using the slice column and also using the slice column and unit column. But in both cases, the presence of many units which do not have any slice in the table does not allow the prefixing of the unit name with the tree branches to be properly done.
https://gist.github.com/devdekunle/b6ac92029a035bd91eea83f8e0edd8a2 
Above is the github gist to see the output of the sorting.
Please can i use another way where i can recursively add units to the table based on the slice name and also create a different table which do not include units without slices so that that table can be generated when the --by-slice mode is passed

poettering commented:
btw: https://systemd.io/CODING_STYLE/

devdekunle commented:
Thank you very much for your review sir @poettering. I will implement the changes ASAP.
